File: light_delivery/light_delivery/report/store_valuation/store_valuation.js
// Copyright (c) 2024, Muhammad Essam and contributors
// For license information, please see license.txt

frappe.query_reports["Store Valuation"] = {
	"filters": [
        {
            "fieldname": "store_name",
            "label": __("Store Name"),
            "fieldtype": "Link",
            "options": "Store",
            "default": "",
            "reqd": 0
        },
        {
            "fieldname": "status",
            "label": __("Status"),
            "fieldtype": "Select",
			"options": "\nPending\nActive\nInactive",
            "reqd": 0
        },
		{
            "fieldname": "zone",
            "label": __("Zone"),
            "fieldtype": "Link",
			"options": "Zone Address",
            "reqd": 0
        },
    ],

	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		// Center the text in specific columns
		if (column.fieldname === "name" || 
			column.fieldname === "store_name" || 
			column.fieldname === "status" || 
			column.fieldname === "store_category" || 
			column.fieldname === "minimum_price" ||
			column.fieldname === "rate_of_km" ||
			column.fieldname === "total_delivered" ||
			column.fieldname === "total_canceled") {
			
			value = `<div style="text-align: center;">${value}</div>`;
		}

		return value;
	}

}

File: light_delivery/light_delivery/report/store_valuation/store_valuation.py
# Copyright (c) 2024, Muhammad Essam and contributors
# For license information, please see license.txt

import frappe
from frappe import _

def get_store_report(filters=None):
    if not filters:
        filters = {}

    stores = frappe.get_all('Store',
        fields=['name', 'store_name', 'status','store_category','minimum_price','rate_of_km'],
        filters=filters
    )

    report_data = []
    
    for store in stores:
        store_name = store['name']
        
        request_data = frappe.db.sql("""
            SELECT
                SUM(CASE WHEN d_r.status = 'Delivered' THEN 1 ELSE 0 END) AS total_delivered,
                SUM(CASE WHEN d_r.status IN ('Delivery Cancel', 'Store Cancel', 'Cancel') THEN 1 ELSE 0 END) AS total_canceled
            FROM `tabRequest Delivery` as d_r
            WHERE store = %s
        """, store_name, as_dict=True)
        
        if request_data:
            store.update({
                'total_delivered': request_data[0].total_delivered,
                'total_canceled': request_data[0].total_canceled
            })

        report_data.append(store)

    return report_data

def get_report_columns():
    return [
        {"label": _("Store ID"), "fieldname": "name", "fieldtype": "Link", "options": "Store", "width": 150},
		{"label": _("Store Name"), "fieldname": "store_name", "fieldtype": "Data", "width": 150},
		{"label": _("Status"), "fieldname": "status", "fieldtype": "Data", "width": 150},
		{"label": _("Store Category"), "fieldname": "store_category", "fieldtype": "Link", "options": "Store Category", "width": 150},
        {"label": _("Minimum Price"), "fieldname": "minimum_price", "fieldtype": "Currency", "width": 140},
        {"label": _("Rate of km"), "fieldname": "rate_of_km", "fieldtype": "Currency", "width": 140},
        {"label": _("Total Requested Delivered"), "fieldname": "total_delivered", "fieldtype": "Int", "width": 120},
        {"label": _("Total Requested Canceled"), "fieldname": "total_canceled", "fieldtype": "Int", "width": 120}
    ]

def execute(filters=None):
	columns = get_report_columns()
	data = get_store_report(filters)
	return columns, data
